### ISSUE TITLE
Swift 3 Support

### ISSUE THREAD

**Issue by rromanchuk** [facebook/facebook-sdk-swift#53]

**Comment by no-donut-for-you** [comment]
+1

**Comment by rromanchuk** [comment]
hey community, is anyone using a known swift3 fork? Has anyone tried running through the conversion tool for this project yet?

**Comment by guitcastro** [comment]
@rromanchuk I saw that you forked the project and created the branch swift3, it's working ?

**Comment by rromanchuk** [comment]
@guitcastro i gave it a try, but couldn't get it to build. I don't know the project very well so I gave up as i'm too busy right now to debug this SDK. I'm just ripping this out for now.

**Comment by gerbiljames** [comment]
I've got a working Swift 3 branch [here](/gerbiljames/facebook-sdk-swift/tree/swift-3.0), it compiles and runs the sample app but Facebook login won't work on iOS 10 (it works fine on iOS 9.3 and lower) due to [this bug](https://developers.facebook.com/bugs/288183311552013/). I don't have a personal Apple developer account to use to enable keychain sharing for the workaround but I guess that would fix the issue on iOS 10.

**Comment by woolfel** [comment]
definitely don't try auto convert, cuz it doesn't build. I just gave it a try since apple pushed out the latest XCode. Tons of compile errors and general breakage.

**Comment by woolfel** [comment]
Nasty surprise for anyone that upgrades XCode to the latest. Even if you don't upgrade (ie convert to swift 3), you won't be able to build your app. I tried setting swift legacy setting in build settings, but the pod doesn't have the setting. Anyone know of a work around for building the pods with legacy swift?

**Comment by woolfel** [comment]
the fix is to add

post_install do |installer|
  installer.pods_project.targets.each do |target|
    target.build_configuration.each do |configuration|
      configuration.build_settings['SWIFT_VERSION'] = "2.3"
    end
  end
end

then run "pods install" to update the workspace. Close the workspace, re-open, clean and build.

**Comment by EvsenevDev** [comment]
what about swift 3.0 and Carthage?
I have error:
`Module compiled with Swift 2.3 cannot be imported in Swift 3.0 `

**Comment by woolfel** [comment]
until Facebook officially supports swift 3.0, I would advise sticking with 2.3 and update the podfile settings. I tried auto-convert and it blew up horribly. That wasted about 4 hours on wednesday that I'll never get back.

**Comment by goufo** [comment]
+1

**Comment by tengpeng** [comment]
@gerbiljames Is your sample app written in Swift 3.0?

**Comment by tengpeng** [comment]
@woolfel I am using Xcode 8 (not beta), and I find the legacy setting in pod building setting.

**Comment by tengpeng** [comment]
@EvsenevDev There is also a linkage error if your pod and your app are of different swift versions.

```
ld: /Users/pt/Library/Developer/Xcode/DerivedData/CurrencyConverter2-diotrdjqfqffqmflbycftzwsojvd/Build/Products/Debug-iphonesimulator/FacebookCore/FacebookCore.framework/FacebookCore compiled with older version of Swift language (2.0) than previous files (3.0) for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)

```

**Comment by woolfel** [comment]
I'm using version 8 (8A218a) to be exact.

1. click on your project
2. change from general to Build settings
3. type in Swift in the search box
4. you should see "swift compiler - Version
5. under that grouping there should be "Use Legacy Swift Language Version"
6. set it to Yes
7. clean and build

That's should set your project to use swift 2.3. You still need to update your podfile to update the facebook sdk to use swift 2.3. If you don't update the podfile and run pods install, the facebook projects won't use swift 2.3 and you'll get build errors. hope that helps

**Comment by EvsenevDev** [comment]
@woolfel with Swift 2.3 all ok. I think we speak about Swift 3.0. Swift 3.0 not support module compiled in Swift 2.3. Faceebook SDK works with Swift 2.3. Project on Swift 3.0 cant work with facebook sdk. For example, if you want use Alamofire => 4.0.0 (support Swift 3.0) and FacebookSDK, your project cant build.

**Comment by gerbiljames** [comment]
@tengpeng It's just the sample app in the fork I linked, translated it to Swift 3 as well when I translated the library.

**Comment by gerbiljames** [comment]
I've got a PR open for Swift 3 support at #55. As of yet doesn't seem to have been looked at.

**Comment by woolfel** [comment]
@EvsenevDev - I didn't have any luck with XCode auto-convert from 2.3 to 3.0. As gerbiljames says, if you really need Swift 3.0 due to other libraries/dependencies, you're out of luck.

My project doesn't use any other libraries that requires 3.0, so I'm ok with 2.3 for now.

**Comment by tengpeng** [comment]
@gerbiljames Found it! I am having trouble getting the Pod work.

My Podfile

```
# Uncomment this line to define a global platform for your project
platform :ios, ’10.0’

target 'Pods/Target Support Files/Pods-SwiftCatalog/Pods-SwiftCatalog.debug.xcconfig' do
  # Uncomment this line if you're using Swift or would like to use dynamic frameworks
  use_frameworks!

  # Pods for SwiftCatalog
pod 'FacebookCore'
pod 'FacebookLogin'
pod 'FacebookShare'

end

```

```
[!] Your Podfile has had smart quotes sanitised. To avoid issues in the future, you should not use TextEdit for editing it. If you are not using TextEdit, you should turn off smart quotes in your editor of choice.

[!] CocoaPods did not set the base configuration of your project because your project already has a custom config set. In order for CocoaPods integration to work at all, please either set the base configurations of the target `SwiftCatalog` to `Pods/Target Support Files/Pods-SwiftCatalog/Pods-SwiftCatalog.debug.xcconfig` or include the `Pods/Target Support Files/Pods-SwiftCatalog/Pods-SwiftCatalog.debug.xcconfig` in your build configuration (`Configurations/SwiftCatalog.xcconfig`).

[!] CocoaPods did not set the base configuration of your project because your project already has a custom config set. In order for CocoaPods integration to work at all, please either set the base configurations of the target `SwiftCatalog` to `Pods/Target Support Files/Pods-SwiftCatalog/Pods-SwiftCatalog.release.xcconfig` or include the `Pods/Target Support Files/Pods-SwiftCatalog/Pods-SwiftCatalog.release.xcconfig` in your build configuration (`Configurations/SwiftCatalog.xcconfig`).
```

**Comment by gerbiljames** [comment]
@tengpeng The sample project in that repo isn't set up for CocoaPods. I've made an example which is [here](https://github.com/gerbiljames/SwiftCatalog).

**Comment by sevenb** [comment]
Unfortunately, I had to install a lower version of Alomofire to work.

source 'https://github.com/CocoaPods/Specs.git'

target 'Test' do
  # Comment this line if you're not using Swift and don't want to use dynamic frameworks
  use_frameworks!

    pod 'Alamofire', '~> 3.0'
    pod 'FacebookCore'
    pod 'FacebookLogin'
    pod 'FacebookShare'

end

post_install do |installer|
    installer.pods_project.targets.each do |target|
        target.build_configurations.each do |configuration|
            configuration.build_settings['SWIFT_VERSION'] = "2.3"
        end
    end
end

**Comment by realmike33** [comment]
I was able to successfully get Facebook to work with Swift 3 by just using their Objective-c pod. You'll also need to change your build options in your project to $(inherited) for Always Embed Swift Standard Libraries and also allow Keychain sharing in the capabilities menu of your xcode project:

`use_frameworks!

platform :ios, '10.0'

def available_pods
    pod 'FBSDKCoreKit', '~> 4.15.1'
    pod 'FBSDKShareKit', '~> 4.15.1'
    pod 'FBSDKLoginKit', '~> 4.15.1'
end
`

**Comment by nlutsenko** [comment]
Hey guys,

Sorry for quite a delay in updating to Swift 3.0, got tangled in few other things/projects.

I am going to take a look at the #55 tomorrow morning in detail and either accept/leave review or, most likely, just re-run the migration and update everything in the repo for Swift 3.0 with a new PR, so we can move fast 😁 

After that, do a round of testing and update the sample app that we've got.
And finally release the Swift 3.0 version.

Long story short: Expect an update sometime next week or so for Swift 3.0.
